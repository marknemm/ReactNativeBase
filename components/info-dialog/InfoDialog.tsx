import { Dialog } from '@rneui/themed';
import type { InfoDialogProps } from './InfoDialog.interfaces';

/**
 * A common dialog that displays information.
 *
 * Contains a title and a close button.
 *
 * @param props The component {@link InfoDialogProps}.
 * @returns The {@link InfoDialog} component.
 */
const InfoDialog: React.FC<InfoDialogProps> = ({
  children,
  isVisible = false,
  onClose,
  title,
  ...dialogProps
}) => (
  <Dialog
    isVisible={isVisible}
    onBackdropPress={onClose}
    {...dialogProps}
  >
    {title && <Dialog.Title title={title} />}
    {children}
    <Dialog.Actions>
      <Dialog.Button onPress={onClose}>Close</Dialog.Button>
    </Dialog.Actions>
  </Dialog>
);

export type * from './InfoDialog.interfaces';
export default InfoDialog;
