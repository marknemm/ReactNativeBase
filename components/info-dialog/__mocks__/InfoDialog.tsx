import type { InfoDialogProps } from '@components/info-dialog/InfoDialog.interfaces';

const MockInfoDialog = jest.fn().mockImplementation((props: InfoDialogProps) => {
  const ActualComponent = jest.requireActual('@components/info-dialog/InfoDialog').default;
  return <ActualComponent {...props} />;
});

export default MockInfoDialog;
