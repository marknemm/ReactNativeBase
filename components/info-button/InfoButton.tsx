import InfoDialog from '@components/info-dialog/InfoDialog';
import { INFO_ICON } from '@constants/icons';
import { Button, useTheme } from '@rneui/themed';
import { forwardRef, useState } from 'react';
import type { InfoButtonProps } from './InfoButton.interfaces';

/**
 * A standard info icon button.
 *
 * Can also display an {@link InfoDialog} on press.
 *
 * @param props The component {@link InfoButtonProps}.
 * @param ref The component reference.
 * @returns The {@link InfoButton} component.
 */
const InfoButton: React.FC<InfoButtonProps> = forwardRef(({
  color = 'transparent',
  iconColor,
  iconSize = 24,
  children,
  dialogTitle,
  onPress,
  ...buttonProps
}, ref) => {
  const { theme } = useTheme();
  const [infoDialogVisible, setInfoDialogVisible] = useState(false);

  return (
    <>
      <Button
        accessibilityLabel="Info"
        color={color}
        icon={{
          ...INFO_ICON,
          color: iconColor ?? theme.colors.black,
          size: iconSize,
        }}
        onPress={(event) => {
          if (children) {
            setInfoDialogVisible(true);
          }
          onPress?.(event);
        }}
        {...buttonProps}
        ref={ref}
      />

      {children && (
        <InfoDialog
          isVisible={infoDialogVisible}
          onClose={() => setInfoDialogVisible(false)}
          title={dialogTitle}
        >
          { children }
        </InfoDialog>
      )}
    </>
  );
});

export type * from './InfoButton.interfaces';
export default InfoButton;
