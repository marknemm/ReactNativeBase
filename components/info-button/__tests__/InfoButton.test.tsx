import InfoButton from '@components/info-button/InfoButton';
import InfoDialog from '@components/info-dialog/InfoDialog';
import { Text } from '@rneui/themed';
import AppProvider from '@test/contexts/app/AppProvider';
import { act, render, screen, userEvent } from '@testing-library/react-native';

jest.mock('@components/info-dialog/InfoDialog');

describe('<InfoButton />', () => {
  const infoDialogStr = 'Info Dialog Test';

  describe('on press', () => {
    it('invokes onPress callback', async () => {
      const onPress = jest.fn();
      render(
        <InfoButton onPress={onPress} />,
        { wrapper: AppProvider }
      );

      const infoButton = screen.getByRole('button', { name: 'Info' });
      await userEvent.press(infoButton);

      expect(onPress).toHaveBeenCalledTimes(1);
    });
  });

  describe('info dialog', () => {
    it('opens the info dialog on press', async () => {
      const onPress = jest.fn();
      render(
        <InfoButton onPress={onPress}>
          <Text>{infoDialogStr}</Text>
        </InfoButton>,
        { wrapper: AppProvider }
      );

      const infoButton = screen.getByRole('button', { name: 'Info' });
      await userEvent.press(infoButton);

      const infoDialog = screen.getByText(infoDialogStr);
      expect(infoDialog).toBeVisible();
      expect(onPress).toHaveBeenCalledTimes(1);
    });

    it('closes the info dialog on close', async () => {
      render(
        <InfoButton>
          <Text>{infoDialogStr}</Text>
        </InfoButton>,
        { wrapper: AppProvider }
      );

      const infoButton = screen.getByRole('button', { name: 'Info' });
      await userEvent.press(infoButton);

      let infoDialog = screen.getByText(infoDialogStr);
      expect(infoDialog).toBeVisible();

      act(() => {
        (InfoDialog as jest.Mock).mock.calls[0][0].onClose();
      });

      infoDialog = screen.queryByText(infoDialogStr);
      expect(infoDialog).toBeFalsy();
    });
  });

  describe('snapshots', () => {
    it('renders correctly', () => {
      render(
        <InfoButton />,
        { wrapper: AppProvider }
      );

      expect(screen.toJSON()).toMatchSnapshot();
    });

    it('renders correctly with info dialog', async () => {
      render(
        <InfoButton>
          <Text>{infoDialogStr}</Text>
        </InfoButton>,
        { wrapper: AppProvider }
      );

      const infoButton = screen.getByRole('button', { name: 'Info' });
      await userEvent.press(infoButton);

      expect(screen.toJSON()).toMatchSnapshot();
    });
  });
});
