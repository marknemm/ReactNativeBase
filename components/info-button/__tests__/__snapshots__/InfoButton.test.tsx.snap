// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InfoButton /> snapshots renders correctly 1`] = `
<View
  style={
    [
      {
        "overflow": "hidden",
      },
      {
        "borderRadius": 3.142857142857143,
      },
      undefined,
      false,
    ]
  }
  testID="RNE_BUTTON_WRAPPER"
>
  <View
    accessibilityLabel="Info"
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": false,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "borderColor": "#007aff",
          "borderRadius": 3.142857142857143,
          "borderWidth": 0,
          "flexDirection": "row",
          "justifyContent": "center",
          "padding": 12.571428571428571,
          "paddingHorizontal": 14.571428571428571,
        }
      }
    >
      <View
        style={
          {
            "marginHorizontal": 5,
            "overflow": "hidden",
          }
        }
        testID="RNE__ICON__CONTAINER"
      >
        <View
          accessibilityRole="button"
          android_ripple={
            {
              "borderless": false,
              "color": "rgba(0, 0, 0, 0.3)",
              "radius": -5,
            }
          }
          disabled={false}
          testID="RNE__ICON__CONTAINER_ACTION"
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "justifyContent": "center",
              }
            }
            testID="RNE__ICON"
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                [
                  {
                    "color": "#242424",
                    "fontSize": 24,
                  },
                  {
                    "backgroundColor": "transparent",
                  },
                  {
                    "fontFamily": "material",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  {},
                ]
              }
              testID="RNE__ICON__Component"
            >
              
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`<InfoButton /> snapshots renders correctly with info dialog 1`] = `
[
  <View
    style={
      [
        {
          "overflow": "hidden",
        },
        {
          "borderRadius": 3.142857142857143,
        },
        undefined,
        false,
      ]
    }
    testID="RNE_BUTTON_WRAPPER"
  >
    <View
      accessibilityLabel="Info"
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": false,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 0.67744,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "borderColor": "#007aff",
            "borderRadius": 3.142857142857143,
            "borderWidth": 0,
            "flexDirection": "row",
            "justifyContent": "center",
            "padding": 12.571428571428571,
            "paddingHorizontal": 14.571428571428571,
          }
        }
      >
        <View
          style={
            {
              "marginHorizontal": 5,
              "overflow": "hidden",
            }
          }
          testID="RNE__ICON__CONTAINER"
        >
          <View
            accessibilityRole="button"
            android_ripple={
              {
                "borderless": false,
                "color": "rgba(0, 0, 0, 0.3)",
                "radius": -5,
              }
            }
            disabled={false}
            testID="RNE__ICON__CONTAINER_ACTION"
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "justifyContent": "center",
                }
              }
              testID="RNE__ICON"
            >
              <Text
                allowFontScaling={false}
                selectable={false}
                style={
                  [
                    {
                      "color": "#242424",
                      "fontSize": 24,
                    },
                    {
                      "backgroundColor": "transparent",
                    },
                    {
                      "fontFamily": "material",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    {},
                  ]
                }
                testID="RNE__ICON__Component"
              >
                
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>,
  <Modal
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    replaceTheme={[Function]}
    testID="Internal__Overlay"
    transparent={true}
    updateTheme={[Function]}
    visible={true}
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "rgba(0, 0, 0, .4)",
          "height": "100%",
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "top": 0,
          "width": "100%",
        }
      }
      testID="RNE__Overlay__backdrop"
    />
    <View
      pointerEvents="box-none"
      style={
        {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
      testID="RNE__Overlay__Container"
    >
      <View
        style={
          {
            "backgroundColor": "rgb(255, 255, 255)",
            "borderRadius": 3,
            "padding": 20,
            "shadowColor": "rgba(0, 0, 0, .3)",
            "shadowOffset": {
              "height": 1,
              "width": 0,
            },
            "shadowRadius": 4,
            "width": "75%",
          }
        }
        testID="RNE__Overlay"
      >
        <Text
          accessibilityRole="text"
          style={
            {
              "color": "#242424",
              "fontFamily": undefined,
              "fontSize": 23.57142857142857,
              "fontStyle": "normal",
              "fontWeight": "normal",
              "marginBottom": 0,
              "textAlign": "left",
              "textDecorationLine": "none",
              "textDecorationStyle": "solid",
            }
          }
        >
          Info Dialog Test
        </Text>
        <View
          style={
            {
              "flexDirection": "row-reverse",
              "flexWrap": "wrap",
              "justifyContent": "flex-start",
              "marginTop": 10,
            }
          }
          testID="Button__View"
        >
          <View
            style={
              [
                {
                  "overflow": "hidden",
                },
                {
                  "borderRadius": 3.142857142857143,
                },
                {
                  "width": "auto",
                },
                false,
              ]
            }
            testID="RNE_BUTTON_WRAPPER"
          >
            <View
              accessibilityRole="button"
              accessibilityState={
                {
                  "busy": false,
                  "checked": undefined,
                  "disabled": false,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "marginLeft": 5,
                  "opacity": 1,
                }
              }
              testID="Dialog__Button"
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "borderColor": "#007aff",
                    "borderRadius": 3.142857142857143,
                    "borderWidth": 0,
                    "flexDirection": "row",
                    "justifyContent": "center",
                    "padding": 12.571428571428571,
                    "paddingHorizontal": 14.571428571428571,
                  }
                }
              >
                <Text
                  style={
                    {
                      "color": "#007aff",
                      "fontSize": 15,
                      "fontWeight": "500",
                      "paddingVertical": 1,
                      "textAlign": "center",
                    }
                  }
                >
                  Close
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </Modal>,
]
`;
